test: Add some unit tests for important internal DHT functions.

iphydf opened the pull request:
We definitely need more of this kind of test so refactorings don't
accidentally break things in ways that happen to still work in auto
tests.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/toktok/c-toxcore/1713)
<!-- Reviewable:end -->

JFreegman commented:
___
*[toxcore/DHT_test.cc, line 44 at r1](https://reviewable.io/reviews/toktok/c-toxcore/1713#-MpmF25s8MfUmL_gt8gR:-MpmF25s8MfUmL_gt8gS:bvccctv) ([raw file](https://github.com/toktok/c-toxcore/blob/e32a501afad660cb15fe9c017b4372db06c4069c/toxcore/DHT_test.cc#L44)):*
> ```cpp
>     random_bytes(pk2.data(), CRYPTO_PUBLIC_KEY_SIZE);
> 
>     if (id_closest(pk0.data(), pk1.data(), pk2.data()) == 0) {
> ```

If two identical keys are ever produced by the RNG in succession we should probably not treat it as expected behaviour. I would either remove that case altogether, or make it abort. Though if you do the latter you're increasing the scope of the test to include an RNG sanity test. Not sure if that's what we want.


<!-- Sent from Reviewable.io -->

iphydf commented:
> If two identical keys are ever produced by the RNG in succession we should probably not treat it as expected behaviour. I would either remove that case altogether, or make it abort. Though if you do the latter you're increasing the scope of the test to include an RNG sanity test. Not sure if that's what we want.

I've added this check. RNG sanity check here isn't too much of a scope increase, it's just making sure that the remainder of the test can rely on some assumptions.

> Edit: is it possible for two different keys to have the same distance? Edit 2: no it's not

It's not. All cryptography relying on xor (which is probably almost all of it) would be broken if it were. I've also added a test to help assert that fact.

JFreegman reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

> Reviewed 3 of 4 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/1713#-:-MpqJHC1ExGIH9cWlNvx:b-jbsf8z)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @iphydf)



<!-- Sent from Reviewable.io -->
